### ISSUE TITLE
Parse Server 3.9 Silent Push Notification on iOS 13 Not Working

### ISSUE THREAD

**Issue by AliZahr** [parse-community/parse-server#6106]
Was anyone able to send silent push notifications? I am unable to receive the push although it says SENT, in the dashboard.

Can anyone send a sample code of the payload if you were able to successfully send one?

**Comment by davimacedo** [comment]
Can you share your current code so we can take a look and try to help you to figure it out? Anyway, I can take a look in [this thread](https://stackoverflow.com/questions/26968902/sending-silent-push-notifications-using-parse-com) as it looks to be a good approach.

**Comment by AliZahr** [comment]
I am sending this push notification from the dashboard with this JSON:
{
"push_type" : "background",
"content-available" : 1,
"priority" : 5,
"videoCallRelated": true,
"videoCallSessionId": "blabla",
"callRejected": true
}
But I am not receiving the silent push notification

**Comment by dplewis** [comment]
Can you use pushType instead of push_type?

https://github.com/parse-community/node-apn/pull/1/files

**Comment by drdaz** [comment]
I'm experiencing a similar issue. And I can see on my device that the notification is being delivered with a priority of 10, despite my having set priority to 5 in the data object.

This same setup (without push-type / pushType and priority) worked fine against iOS 12.

**Comment by drdaz** [comment]
Yeah something seems wrong here. Here's a snippet from an afterSave method I've got:

```
    const data = {

        where: installationQuery,
        
        data: {
            'content-available': 1,
            'category': 'MessageCategoryId',
            'messageId': objectId,
            'apns-priority': 5,
            'pushType': 'background'
        }
    }

    Parse.Push.send(data, { useMasterKey: true }).then(function() {
        // Successful Push
        console.log("Push success!")
    }, function(error) {
        // Failed Push
        console.log("Push failed :(")
    })
```

When one of these notifications is received on the device, this is logged in the console:

```
default	22:44:09.091864 +0200	SpringBoard	Received incoming message on topic org.example.myApp at priority 10

default	22:44:09.095459 +0200	SpringBoard	[org.example.myApp] Received remote notification request F694-86D3 [ hasAlertContent: 0, hasSound: 0 hasBadge: 0 hasContentAvailable: 1 hasMutableContent: 0 pushType: Alert]
```

So both the priority and the push type are wrong. Any ideas what I'm doing wrong here?

**Comment by drdaz** [comment]
I've spent most of the day trying to figure out how to debug installed modules in node; I'm an iOS guy usually. 

Now that I've got there, the implementation of the new iOS flags seems off. 

In node_apn/lib/notification/index.js in the exported `headers`method, where we check for, for example, `this.pushType` before setting the value on the notification. But the pushType value we set in 'data' is in `this.payload.pushType` at this point.

It's the same mistake for all the others it seems.

The tests in node-apn seem to test the default values given to the notification at init (like priority===10 and pushType===alert). That's why they pass.

My notifications have started working again locally by adding .payload in all the appropriate places in that `headers` method.

I'll create a PR; this seems broken.

**Comment by drdaz** [comment]
@davimacedo Can you comment? I see you've been active on that repo.

**Comment by dplewis** [comment]
@drdaz Please check the [Push Notification Guide](https://docs.parseplatform.org/parse-server/guide/#api)

```
data: {
            'content-available': 1,
            'category': 'MessageCategoryId',
            'messageId': objectId,
            'apns-priority': 5, // should be priority
            'pushType': 'background' // should be push_type
}
```

**Comment by drdaz** [comment]
@dplewis I'm aware what the documentation says. What I posted here was a snippet where I was experimenting.

The documentation is wrong about pushType, and the headers aren't being sent.

**Comment by dplewis** [comment]
@drdaz @acinader @davimacedo 

Looking at the [JS Docs](https://docs.parseplatform.org/js/guide/#customizing-your-notifications) headers like` push_time` and `priority` should be in the data field. There are other headers like `expiration_time` aren't in the data field.

The push adapter doesn't check the data field for headers. I want to discuss what we should do going forward.

Two solutions:
A: SDK update
B: Push Adapter Update to support headers in data field 
https://github.com/parse-community/parse-server-push-adapter/blob/master/src/APNS.js#L76
`const pushType = data['push_type'] || coreData['push_type']`

```
Parse.Push.send({
  channels: [ "Mets" ],
  expiration_time: oneWeekAway, // Header
  push_time: ... // This will require an SDK / server update to work, push adapter is expecting this.
  data: {
    alert: "The Mets scored! The game is now tied 1-1.",
    badge: "Increment",
    sound: "cheering.caf",
    title: "Mets Score!",
    push_type: ... // No SDK update required, not supported by push adapter, is in documentation
  }
})
```

Thoughts?

**Comment by dplewis** [comment]
I updated my comment

**Comment by drdaz** [comment]
I tried this adding `push_type: 'background'` and `priority: 5` in the root of the object like you describe here, and that does seem to work. Although the console output on the device is a little odd; the initial message suggests that the notification is sent at priority 1, but delivered to the app with priority 5:

```
default	10:56:55.382793 +0200	SpringBoard	Received incoming message on topic org.example.myApp at priority 1
default	10:56:55.386442 +0200	SpringBoard	[org.example.myApp] Received remote notification request CAC9-8138 [ hasAlertContent: 0, hasSound: 0 hasBadge: 0 hasContentAvailable: 1 hasMutableContent: 0 pushType: Background]
default	10:56:55.386517 +0200	SpringBoard	[org.example.myApp] Deliver push notification CAC9-8138
default	10:56:55.386594 +0200	SpringBoard	[org.example.myApp] Request DUET delivers content-available push notification to application
default	10:56:55.387768 +0200	dasd	Submitted Activity: com.apple.pushLaunch.org.example.myApp:B5AA90 at priority 5 <private>

```

I don't *think* that was happening with the mods I made in the other PR.

Just tested it. Setting priority to a value other than 5 doesn't work. The incoming priority value on the device is always 1, and the notification is always delivered to the app with priority 5 as shown in that log snippet.

Oddly enough, the push_type value seems to get sent just fine.

**Comment by drdaz** [comment]
TL;DR Yeah, that works. 😂

**Comment by MohammadAdib** [comment]
Are any of you guys able to send background notifications? I only receive them in-app on iOS...

Here is the payload I am using

`var data = {
    'aps': {
      'content-available': 1,
      'apns-priority': 5,
      'push_type': 'background',
      'notificationId': id,
      'sound': ''
    },
    'content-available': 1,
    'notificationId': id,
    'push_type': 'background'
  };
`

And I'm sending it using this code:

`await Parse.Push.send({
        channels: [channel],
        push_type: 'background',
        priority: 5,
        data: data
      }, { useMasterKey: true });`

**Comment by dplewis** [comment]
Read my https://github.com/parse-community/parse-server/issues/6106#issuecomment-544204292

This will be fixed in the next release. You don’t add push_type in the data field

**Comment by MohammadAdib** [comment]
> Read my [#6106 (comment)](https://github.com/parse-community/parse-server/issues/6106#issuecomment-544204292)
> 
> This will be fixed in the next release. You don’t add push_type in the data field.
> 
> Edit: sorry just read it, if you are able to receive it in your app is there an issue?

I am trying to receive the notification in the background even if the app is in the background. So if the app is not in foreground, I still want to receive a silent push. Is this possible on iOS 13? It has been working for a long time. This is for the app CarMeets. www.carmeets.app

**Comment by dplewis** [comment]
https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_background_updates_to_your_app

Background apps have a 30 second max before they sleep.

https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_background_updates_to_your_app

From what I read here, a background app will wake up for 30 seconds from a silent push. 

I may have misread it but feel free to open an issue one on the node-apn repo

**Comment by dplewis** [comment]
Closing via [3.10.0](https://github.com/parse-community/parse-server/pull/6321)
